### PR TITLE
Update worldedit-sponge to API 8

### PR DESCRIPTION
**Pull request by octylFractal**
This replaces #1289 and updates to the current tip, along with incorporating #1766 for a better lifecycle. It's currently non-functional due to what I presume is Sponge bugs, but once it's all ironed out and confirmed to be working, it'll become ready for review! Feel free to take a look now.

- [x] Fix completions
- [x] Re-add bStats (https://github.com/Bastian/bStats-Metrics/issues/97)
- [x] Investigate VALIDATION side effect

### CODE REVIEW

**Comment by me4502** [comment]
Partially blocked by https://github.com/SpongePowered/Sponge/issues/3585 (Unsure if acceptable to merge before this is fixed)

**Review by me4502 — APPROVED** [review]
From my look-overs and testing this appears to be good, aside from otherwise noted external issues

File: worldedit-sponge/build.gradle.kts [hunk 1]
```diff
@@ -1,54 +1,87 @@
...
-    maven { url = uri("https://repo.codemc.org/repository/maven-public") }
+minecraft {
+    version("1.16.5")
+}
+
+val spongeApiVersion = "8.0.0";
+
+sponge {
+    apiVersion(spongeApiVersion)
+    license("GPL3")
```
Comment (octylFractal) [inline worldedit-sponge/build.gradle.kts:25]:
```suggestion
    license("GPL-3.0-or-later")
```
it ain't officially spdx but that doesn't mean we can't do it

File: worldedit-sponge/build.gradle.kts [hunk 2]
```diff
@@ -1,54 +1,87 @@
...
+        }
+    }
 }
 
 dependencies {
     api(project(":worldedit-core"))
     api(project(":worldedit-libs:sponge"))
-    api("org.spongepowered:spongeapi:7.1.0") {
+    api("org.spongepowered:spongeapi:${spongeApiVersion}") {
         exclude(group = "org.slf4j", module = "slf4j-api")
```
Comment (octylFractal) [inline worldedit-sponge/build.gradle.kts:54]: This is probably not needed anymore? IDK.

File: worldedit-sponge/build.gradle.kts [hunk 3]
```diff
@@ -1,54 +1,87 @@
...
     }
     implementation(platform("org.apache.logging.log4j:log4j-bom:${Versions.LOG4J}") {
         because("Sponge 8 (will?) provides Log4J")
     })
     api("org.apache.logging.log4j:log4j-api")
-    api("org.bstats:bstats-sponge:1.7")
-    testImplementation("org.mockito:mockito-core:1.9.0-rc1")
+    // bStats isn't updated yet :(
+    api("org.bstats:bstats-sponge:2.2.1")
```
Comment (octylFractal) [inline worldedit-sponge/build.gradle.kts:61]: if it's not updated why is it here? shouldn't it be commented out or removed altogether?

File: worldedit-sponge/build.gradle.kts [hunk 4]
```diff
@@ -1,54 +1,87 @@
...
 
-    configure<SigningExtension> {
-        sign("shadowJar")
-    }
-
-    tasks.named("build").configure {
-        dependsOn("signShadowJar")
+        relocate("org.antlr.v4", "com.sk89q.worldedit.sponge.antlr4")
+        include(dependency("org.antlr:antlr4-runtime"))
+        relocate("it.unimi.dsi.fastutil", "com.sk89q.worldedit.sponge.fastutil") {
```
Comment (octylFractal) [inline worldedit-sponge/build.gradle.kts:80]: Does sponge really not provide fastutil?

File: worldedit-sponge/build.gradle.kts (hunk 2 above)
Comment (zml2008) [reply worldedit-sponge/build.gradle.kts:54]: You likely don't need to declare a SpongeAPI dependency at all, we add it to your compile classpath based on the version you declare in the SpongeGradle extension

File: worldedit-sponge/build.gradle.kts (hunk 4 above)
Comment (zml2008) [reply worldedit-sponge/build.gradle.kts:80]: Would you mind opening an issue for this?

Comment (octylFractal) [reply worldedit-sponge/build.gradle.kts:76]: Was done in https://github.com/SpongePowered/SpongeAPI/issues/2408.

File: worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongePlatform.java [hunk 5]
```diff
@@ -197,15 +177,13 @@ public String getPlatformVersion() {
 
     @Override
     public Set<SideEffect> getSupportedSideEffects() {
-        return ImmutableSet.of();
+        return ImmutableSet.of(
+            SideEffect.UPDATE, SideEffect.ENTITY_AI, SideEffect.LIGHTING, SideEffect.NEIGHBORS
```
Comment (octylFractal) [inline worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongePlatform.java:181]: No `VALIDATION`?

File: worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongeWorld.java [hunk 6]
```diff
@@ -19,204 +19,358 @@
...
+                .withUpdateNeighbors(sideEffects.shouldApply(SideEffect.NEIGHBORS))
+                .withNotifyClients(true)
+                .withPhysics(sideEffects.shouldApply(SideEffect.UPDATE))
+                .withNotifyObservers(sideEffects.shouldApply(SideEffect.UPDATE))
+                .withLightingUpdates(sideEffects.shouldApply(SideEffect.LIGHTING))
+                .withPathfindingUpdates(sideEffects.shouldApply(SideEffect.ENTITY_AI))
+                .withNeighborDropsAllowed(false)
+                .withBlocksMoving(false)
+                .withForcedReRender(false)
+                .withIgnoreRender(false)
```
Comment (octylFractal) [inline worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongeWorld.java:184]: I wonder if this should be cached -- worth measuring performance difference if you use e.g. an `EnumMap` to store pre-calculated flags.

File: worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongePlatform.java (hunk 5 above)
Comment (me4502) [reply worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongePlatform.java:181]: No, Sponge does not support it as a separate side effect

**Pull request merged**